File: src/app/conjugation/forms/potential-affirmative.ts
import {Conjugation, Word, WordType} from "../conjugation";
import {HIRAGANA} from "../hiragana";

export class PotentialAffirmative implements Conjugation {
    getConjugation(word: Word): (Word | undefined) {
        switch (word.wordType) {

            // Verbs
            case WordType.GodanVerb:
                const lastKana = HIRAGANA[word.getLastKana()]

                return word.replaceLastKana(lastKana.getGroup().e + 'る')

            case WordType.IchidanVerb:
                return word.replaceLastKana('られる');

            case WordType.SuruVerb:
                return word.replaceLastKana('ができる', 2);

            case WordType.KuruVerb:
                return word.replace('来られる', 'こられる');


            default:
                return undefined;
        }
    }

    getTitle(): string {
        return "Potential, affirmative"
    }

    getSettingsTitle(): string {
        return "Affirmative"
    }

}

File: backend/app/conjugation/forms/potential_affirmative.py
from ..core import Conjugation, Word, WordType
from ..hiragana import HIRAGANA


class PotentialAffirmative(Conjugation):

    title = 'Potential, affirmative'
    settings_title = 'Affirmative'

    def conjugate(self, word: Word) -> Word | None:
        match word.word_type:

            # Verbs
            case WordType.GODAN_VERB:
                last_kana = HIRAGANA[word.last_kana]

                return word.replace_last_kana(last_kana.group.e + 'る')

            case WordType.ICHIDAN_VERB:
                return word.replace_last_kana('られる')

            case WordType.SURU_VERB:
                return word.replace_last_kana('ができる', 2)

            case WordType.KURU_VERB:
                return word.replace('来られる', 'こられる')

            case _:
                return None
